js/src/Components/Behaviors.js

```diff
@@ -0,0 +1,188 @@
+    }
+
+    onBehaviorChange(itemId, entityType, behaviorKey, fieldName, behaviorValues)
+  }
+
+
+  /**
+   * Flatten a multidimensional object
+   *
+   * For example:
+   *   flattenObject({ a: 1, b: { c: 2 } })
+   * Returns:
+   *   { a: 1, c: 2}
+   */
+  const flattenObject = (obj) => {
+    const flattened = {}
+
+    Object.keys(obj).forEach((key) => {
+      if (typeof obj[key] === 'object' && obj[key] !== null) {
+        Object.assign(flattened, flattenObject(obj[key]))
+      }
+      else {
+        flattened[key] = obj[key]
+      }
+    })
+
+    return flattened
+  }
+
+  /**
+   * Get the field widget settings object.
+   *
+   * Since we are using the widgets similar to the regular content fields, we
+   * need build an object with some settings.
+   *
+   * @param field
+   * @returns {{}}
+   */
+  const getFieldSettings = (field) => {
+    const settings = {};
+    Object.keys(field).map(key => {
+      const newKey = key.replace('#', '');
+      settings[newKey] = field[key];
+    })
+    settings.label = settings.title;
+    settings.help = settings.description;
+    settings.column_key = 'value';
+    settings.cardinality = (typeof settings.multiple !== 'undefined' && settings.multiple) ? -1 : 1;
+
+    if (settings.type === 'checkbox') {
+      settings.options = {1: settings.title};
+    }
+    return settings;
+  }
+
+  /**
+   * Get the default value for the field.
+   *
+   * @param behaviorKey
+   * @param fieldName
+   * @returns {{value}[]|*[]|*}
+   */
+  const getDefaultBehaviorValue = (behaviorKey, fieldName) => {
+    let default_value;
+
+    if (
+      typeof entity.behavior_settings === 'undefined' ||
+      typeof entity.behavior_settings[0].value[behaviorKey] === 'undefined' ||
+      typeof entity.behavior_settings[0].value[behaviorKey][fieldName] === 'undefined'
+    ) {
+      default_value = typeof behaviors[behaviorKey][fieldName]['#default_value'] === 'undefined' ? [] : behaviors[behaviorKey][fieldName]['#default_value'];
+    }
+    else {
+      default_value = entity.behavior_settings[0].value[behaviorKey][fieldName];
+    }
+
+    switch (behaviors[behaviorKey][fieldName]['#type']) {
+      case 'checkbox':
+        // Convert true/false into integers for the field widget.
+        default_value = default_value ? 1 : 0;
+        break;
+
+      case 'checkboxes':
+        // Construct an array of values to match what the field widget wants.
+        return default_value.map(item => {
+          return ({value: item})
+        });
+
+      case 'select':
+        // Construct an array of values to match what the field widget wants if
+        // the select field allows multiple values.
+        if (behaviors[behaviorKey][fieldName]['#multiple']) {
+          return default_value.map(item => {
+            return ({value: item})
+          });
+        }
+        break;
+    }
+
+    return [{value: default_value}];
+  }
+
+  return (
+    <div>
+      {Object.keys(behaviors).map(behaviorKey =>
```

Comment:
As a style convention in the future for consideration, when you get multiple or big map functions it seems better to pull it up out of the return jsx statement and make them their own variable or function.

Reply:
To me that makes tracing the render function more difficult because we have to go through multiple different functions to just see the markup structure. But there's probably a better way to do this still.